controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+    MessageRequiresData,
+
+    #[error("A serialization error occurred: {0}")]
+    SerDes(hubpack::Error),
+}
+
+// A request sent from host to SP, possibly with trailing data.
+#[derive(Clone, Debug)]
+struct HostRpcRequest {
+    pub message: Message,
+    pub data: Option<Vec<u8>>,
+}
+
+// A response sent from SP to host, possibly with trailing data.
+#[derive(Clone, Debug)]
+#[allow(dead_code)]
+struct SpRpcResponse {
+    pub message: Message,
+    pub data: Option<Vec<u8>>,
+}
+
+/// A request sent from SP to host, possibly with trailing data.
+#[derive(Clone, Debug)]
+pub struct SpRpcRequest {
+    pub message: Message,
+    pub data: Option<Vec<u8>>,
+}
+
+/// A response sent from host to SP, possibly with trailing data.
+#[derive(Clone, Debug)]
+pub struct HostRpcResponse {
+    pub message: Message,
+    pub data: Option<Vec<u8>>,
+}
+
+// A request from host to SP that has not yet been completed.
+#[derive(Debug)]
+struct OutstandingHostRequest {
+    // The actual request object we're sending. It's stored so that we can
+    // resend it if needed.
+    request: HostRpcRequest,
+    // The channel on which the eventual reply will be sent.
+    response_tx: oneshot::Sender<Result<SpRpcResponse, Error>>,
+    // The time we last sent the request, used during retries.
+    last_sent: Instant,
+}
+
+// We limit ourselves to a single outstanding request in either direction at
+// this point.
+const NUM_OUTSTANDING_REQUESTS: usize = 1;
 
 /// A type for controlling transceiver modules on a Sidecar.
 #[derive(Debug)]
-pub struct Controller;
+pub struct Controller {
+    _log: Logger,
+    message_id: AtomicU64,
+
+    // Channel onto which requests from the host to SP are sent.
+    //
+    // `io_task` owns the receiving end of this, and actually sends out the
+    // messages to the SP.
+    outgoing_request_tx: mpsc::Sender<OutstandingHostRequest>,
+
+    // The task handling actual network IO with the peer.
+    io_task: JoinHandle<()>,
+
+    // The task handling actual network IO with the peer.
+    request_task: JoinHandle<()>,
+}
+
+impl Drop for Controller {
+    fn drop(&mut self) {
+        self.io_task.abort();
+        self.request_task.abort();
+    }
+}
+
+impl Controller {
+    /// Create a new transceiver controller.
+    ///
+    /// `request_handler` is a function that yields responses to SP requests. As
+    /// requests over the network are received, they'll be passed into the
+    /// handler, and the yielded response forwarded back to the SP.
+    pub async fn new<H, F>(log: Logger, request_handler: H) -> Result<Self, Error>
+    where
+        H: Fn(SpRpcRequest) -> F + Send + Sync + 'static,
+        F: Future<Output = Result<HostRpcResponse, Error>> + Send,
+    {
+        // TODO-correctness We probably want to accept a specific address as
+        // part of the construction of this object.
+        let local_addr = SocketAddrV6::new(Ipv6Addr::UNSPECIFIED, PORT, 0, 0);
```

Comment:
May not be needed on this PR, but we will almost certainly need to set the `scope_id` here and in `join_multicast_v6()` below based on the interface that this socket should use. The way MGS does this currently (which may not be right long term) is to wait for the interface to exist, determined by calling `if_nametoindex()` in a loop until it succeeds: https://github.com/oxidecomputer/management-gateway-service/blob/d7791e4d02571b8414b4c2976b0439234aa4d8c7/gateway-sp-comms/src/single_sp/startup.rs#L219-L246

Reply:
Agreed, and I'm also hoping to ignore this right now. (Though I should make an issue to track.)

Just thinking out loud for a minute: thi type will be consumed by `dpd`, the main Dendrite binary. That will run inside the switch zone, launched by the sled agent (bootstrap agent, really). All of the zones the sled agent currently kicks off have a VNIC over the inter-zone etherstub, along with an IPv6 link-local address. The interface name could be provided to `dpd`, just like it is for MGS. So `if_nametoindex` seems like the right answer here.

Something about the retry loop seems odd, since if it doesn't exist, then the sled agent didn't start things up correctly. But as you pointed out, that's a longer term concern! Thanks.

Reply:
See https://github.com/oxidecomputer/transceiver-control/issues/14